File: src/test/csrc/common/sdcard.cpp
/***************************************************************************************
* Copyright (c) 2020-2023 Institute of Computing Technology, Chinese Academy of Sciences
* Copyright (c) 2020-2021 Peng Cheng Laboratory
*
* DiffTest is licensed under Mulan PSL v2.
* You can use this software according to the terms and conditions of the Mulan PSL v2.
* You may obtain a copy of Mulan PSL v2 at:
*          http://license.coscl.org.cn/MulanPSL2
*
* THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
* EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
* MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
*
* See the Mulan PSL v2 for more details.
***************************************************************************************/

#include "sdcard.h"
#include "common.h"
#ifdef CONFIG_DIFFTEST_PERFCNT
#include "perf.h"
#endif // CONFIG_DIFFTEST_PERFCNT

FILE *fp = NULL;

void check_sdcard() {
  if (!fp) {
    eprintf(ANSI_COLOR_MAGENTA "[warning] sdcard img not found\n");
  }
}

void sd_setaddr(uint32_t addr) {
#ifdef CONFIG_DIFFTEST_PERFCNT
  difftest_calls[perf_sd_set_addr]++;
  difftest_bytes[perf_sd_set_addr] += 4;
#endif // CONFIG_DIFFTEST_PERFCNT
  check_sdcard();
#ifdef SDCARD_IMAGE
  fseek(fp, addr, SEEK_SET);
#endif
  //printf("set addr to 0x%08x\n", addr);
  //assert(0);
}

void sd_read(uint32_t *data) {
#ifdef CONFIG_DIFFTEST_PERFCNT
  difftest_calls[perf_sd_read]++;
  difftest_bytes[perf_sd_read] += 4;
#endif // CONFIG_DIFFTEST_PERFCNT
  check_sdcard();
#ifdef SDCARD_IMAGE
  fread(data, 4, 1, fp);
#endif
  //printf("read data = 0x%08x\n", *data);
  //assert(0);
}

void init_sd(void) {
#ifdef SDCARD_IMAGE
  fp = fopen(SDCARD_IMAGE, "r");
  check_sdcard();
#endif
}

void finish_sd(void) {
#ifdef SDCARD_IMAGE
  fclose(fp);
#endif
}

File: src/main/scala/common/SDCard.scala
/***************************************************************************************
* Copyright (c) 2020-2024 Institute of Computing Technology, Chinese Academy of Sciences
*
* XiangShan is licensed under Mulan PSL v2.
* You can use this software according to the terms and conditions of the Mulan PSL v2.
* You may obtain a copy of Mulan PSL v2 at:
*          http://license.coscl.org.cn/MulanPSL2
*
* THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
* EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
* MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
*
* See the Mulan PSL v2 for more details.
***************************************************************************************/

package difftest.common

import chisel3._
import chisel3.experimental.ExtModule
import chisel3.util._

class DifftestSDCardRead extends Bundle {
  val setAddr = Input(Bool())
  val addr = Input(UInt(32.W))
  val ren = Input(Bool())
  val data = Output(UInt(32.W))
}

class SDCardHelper extends ExtModule with HasExtModuleInline {
  val clock = IO(Input(Clock()))
  val io = IO(new DifftestSDCardRead)

  val cppExtModule =
    """
      |void SDCardHelper (
      |  uint8_t   io_setAddr,
      |  uint32_t  io_addr,
      |  uint8_t   io_ren,
      |  uint32_t& io_data
      |) {
      |  if (io_ren) sd_read(&io_data);
      |  if (io_setAddr) sd_setaddr(io_addr);
      |}
      |""".stripMargin
  difftest.DifftestModule.createCppExtModule("SDCardHelper", cppExtModule, Some("\"sdcard.h\""))

  setInline(
    "SDCardHelper.v",
    s"""
       |`ifndef SYNTHESIS
       |import "DPI-C" function void sd_setaddr(input int addr);
       |import "DPI-C" function void sd_read(output int data);
       |`endif // SYNTHESIS
       |
       |module SDCardHelper (
       |  input clock,
       |  input io_setAddr,
       |  input [31:0] io_addr,
       |  input io_ren,
       |  output reg [31:0] io_data
       |);
       |
       |`ifndef SYNTHESIS
       |  always@(negedge clock) begin
       |    if (io_ren) sd_read(io_data);
       |  end
       |  always@(posedge clock) begin
       |    if (io_setAddr) sd_setaddr(io_addr);
       |  end
       |`endif // SYNTHESIS
       |
       |endmodule
     """.stripMargin,
  )
}

class DifftestSDCard extends Module {
  val io = IO(new DifftestSDCardRead)

  val helper = Module(new SDCardHelper)
  helper.clock := clock

  io <> helper.io
}

object DifftestSDCard {
  def apply(): DifftestSDCardRead = {
    Module(new DifftestSDCard).io
  }
}
